datalad/interface/tests/test_utils.py

```diff
@@ -0,0 +1,82 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Test dirty dataset handling
+
+"""
+
+__docformat__ = 'restructuredtext'
+
+from os.path import join as opj
+from nose.tools import assert_raises, assert_equal
+from datalad.tests.utils import with_tempfile, assert_not_equal
+from datalad.interface.utils import handle_dirty_dataset
+from datalad.distribution.dataset import Dataset
+
+_dirty_modes = ('fail', 'ignore', 'save-before')
+
+
+def _check_all_clean(ds, state):
+    assert state is not None
+    for mode in _dirty_modes:
+        # nothing wrong, nothing saved
+        handle_dirty_dataset(ds, mode)
+        assert_equal(state, ds.repo.get_hexsha())
+
+
+def _check_auto_save(ds, orig_state):
+    handle_dirty_dataset(ds, 'ignore')
+    assert_raises(RuntimeError, handle_dirty_dataset, ds, 'fail')
+    handle_dirty_dataset(ds, 'save-before')
+    state = ds.repo.get_hexsha()
+    assert_not_equal(orig_state, state)
+    _check_all_clean(ds, state)
+    return state
+
+
+@with_tempfile(mkdir=True)
+def test_dirty(path):
+    for mode in _dirty_modes:
+        # does nothing without a dataset
+        handle_dirty_dataset(None, mode)
+    # placeholder, but not yet created
+    ds = Dataset(path)
+    # unknown mode
+    assert_raises(ValueError, handle_dirty_dataset, ds, 'MADEUP')
+    # not yet created is very dirty
+    assert_raises(RuntimeError, handle_dirty_dataset, ds, 'fail')
+    handle_dirty_dataset(ds, 'ignore')
+    assert_raises(RuntimeError, handle_dirty_dataset, ds, 'save-before')
+    # should yield a clean repo
+    ds.create()
+    orig_state = ds.repo.get_hexsha()
+    _check_all_clean(ds, orig_state)
+    # tainted: untracked
+    with open(opj(ds.path, 'something'), 'w') as f:
+        f.write('some')
+    orig_state = _check_auto_save(ds, orig_state)
+    # tainted: staged
+    with open(opj(ds.path, 'staged'), 'w') as f:
+        f.write('some')
+    ds.repo.add('staged', git=True)
+    orig_state = _check_auto_save(ds, orig_state)
+    # tainted: submodule
+    # not added to super on purpose!
+    subds = Dataset(opj(ds.path, 'subds')).create(add_to_super=False)
+    _check_all_clean(subds, subds.repo.get_hexsha())
+    orig_state = _check_auto_save(ds, orig_state)
+    # XXX surprisingly this is added as a submodule, but there is no .gitmodules
+    # which confused even Git itself (git submodule call now fails with
+    # "fatal: no submodule mapping found in .gitmodules for path 'subds'"
+    assert_equal(ds.get_subdatasets(), [])
```

Comment:
from a brief chat -- I don't think that anything would be done on git side

that "Subproject" commit is added by 'git add', and kinda "independent" from all the submodule mechanism, but submodule freaks out when sees Subproject without mapping (as we discovered).  As to me -- awkward design... but oh well -- we can easily workaround, so we should.

Also (re)discovered the 'subtree' mechanism. Verified that it doesn't deal in terms of "Subproject" commits -- really does merges and records state information within the commits seems to me, e.g.:
```
commit 7faf2ffe6661f6957e48e104d7e20a83041aba1c
Merge: cf74de7 6210ac5
Author: Yaroslav Halchenko <debian@onerussian.com>
Date:   Fri Sep 2 14:35:51 2016 -0400

    Add 'tools/' from commit '6210ac540b07dd094004cdab6fce2ce9baf636fc'
    
    git-subtree-dir: tools
    git-subtree-mainline: cf74de722077e1ca8814e0ec9a6c893d7dbdda8d
    git-subtree-split: 6210ac540b07dd094004cdab6fce2ce9baf636fc
```
whenever I split `tools/` into another branch, removed in master, and then 'subtree add'ed from the branch it was split into.  Just beware of the beast ;)